worker/remoterelations/remoterelations.go

```diff
@@ -388,86 +369,117 @@ func (w *remoteApplicationWorker) relationChanged(
+		w.relationInfo.remoteEndpointName = remoteRelation.RemoteEndpointName
+		var err error
+		applicationTag := names.NewApplicationTag(remoteRelation.LocalApplicationName)
+		w.relationInfo.applicationId, remoteRelationId, err = w.registerRemoteRelation(applicationTag, relationTag, publisher)
+		if err != nil {
 			return errors.Trace(err)
 		}
-		r := &relation{}
-		r.RelationId = remoteRelation.Id
-		r.Life = remoteRelation.Life
-		r.ruw = relationUnitsWatcher
-		relations[key] = r
 	}
+
+	// Start a watcher to track changes to the local units in the
+	// relation, and a worker to process those changes.
+	localRelationUnitsWatcher, err := w.facade.WatchLocalRelationUnits(key)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	relationUnitsWatcher, err := newRelationUnitsWatcher(
+		relationTag,
+		w.relationInfo.applicationId,
+		remoteRelationId,
+		localRelationUnitsWatcher,
+		w.relationChanges,
+		w.facade,
+	)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := w.catacomb.Add(relationUnitsWatcher); err != nil {
+		return errors.Trace(err)
+	}
+	r := &relation{}
+	r.RelationId = remoteRelation.Id
+	r.Life = remoteRelation.Life
+	r.ruw = relationUnitsWatcher
+	relations[key] = r
 	return nil
 }
 
 func (w *remoteApplicationWorker) registerRemoteRelation(
-	relationTag names.Tag, publisher RemoteRelationChangePublisher,
-) (params.RemoteEntityId, error) {
+	applicationTag, relationTag names.Tag, publisher RemoteRelationChangePublisher,
+) (remoteApplicationId params.RemoteEntityId, remoteRelationId params.RemoteEntityId, _ error) {
 	logger.Debugf("register remote relation %v", relationTag.Id())
+	emptyId := params.RemoteEntityId{}
 	// Ensure the relation is exported first up.
-	results, err := w.facade.ExportEntities([]names.Tag{relationTag})
+	results, err := w.facade.ExportEntities([]names.Tag{applicationTag, relationTag})
 	if err != nil {
-		return params.RemoteEntityId{}, errors.Annotatef(err, "exporting relation %v", relationTag)
+		return emptyId, emptyId, errors.Annotatef(err, "exporting relation %v and application", relationTag, applicationTag)
 	}
 	if results[0].Error != nil && !params.IsCodeAlreadyExists(results[0].Error) {
-		return params.RemoteEntityId{}, errors.Annotatef(err, "exporting relation %v", relationTag)
+		return emptyId, emptyId, errors.Annotatef(err, "exporting application %v", applicationTag)
+	}
+	remoteApplicationId = *results[0].Result
+	if results[1].Error != nil && !params.IsCodeAlreadyExists(results[1].Error) {
+		return emptyId, emptyId, errors.Annotatef(err, "exporting relation %v", relationTag)
 	}
-	remoteRelationId := *results[0].Result
+	remoteRelationId = *results[1].Result
 
 	// This data goes to the remote model so we map local info
 	// from this model to the remote arg vales and visa versa.
 	arg := params.RegisterRemoteRelation{
-		ApplicationId:          w.relationInfo.applicationId,
+		ApplicationId:          remoteApplicationId,
```

Comment:
And this is the bit where it was sending the token for the wrong side, is that right?